### ISSUE TITLE
Fix coarse modules that reveals the complex code dependencies

### ISSUE THREAD

**Issue by wangkuiyi** [PaddlePaddle/Paddle#763]
`cmake` doesn't require explicit descriptions of fine-grained code dependencies -- modules are often directories and dependencies are defined between these directories.  This discourages developers to think and describe code dependencies clearly and leads to cyclic #include dependencies without and between directories.

For example, here is the #include dependencies within directory `/paddle/cuda/include`:

![a](https://cloud.githubusercontent.com/assets/1548775/20981817/ed9deba6-bc6a-11e6-88ed-6d928a049c95.png)

We can see cycles in the figure.

Also, `paddle/cuda` and `paddle/math` depend on each other.  The following files in `/paddle/cuda` includes header files in `paddle/math`:

```
hl_cpu_gru.cuh:#include "paddle/math/MathFunctions.h"
```

and these files in `/paddle/math` includes files in `/paddle/cuda`:

```
Allocator.h:#include "hl_gpu.h"
BaseMatrix.cu:#include "hl_matrix_ops.cuh"
```

`paddle/utils` and `paddle/cuda` depend on each other too.

**Comment by wangkuiyi** [comment]
Recently I changed our Dockerfile to generate `/paddle/build/compile_commands.json` which records all building commands executed by `cmake`.  I did this so that `woboq code browser` can use it to understand our C++ code and generated browsable version in HTML files.  But I noticed that this also helps us understand the current `cmake`-based building process and code dependencies.

In order to generate `/paddle/build/compile_commands.json`, we need to build a Docker image from the most recent source code:

```
cd ~/work
git clone --recursive https://github.com/paddlepaddle/paddle
cd paddle
docker build -t paddle:gpu -f paddle/scripts/docker/Dockerfile.gpu .
```

Once this is done, we can run the Docker image and a container:

```
docker run -d -p 2022:22 --name paddle-gpu-doc paddle:gpu
```

Because the Docker image includes an SSHD server, so we can use `scp` to retrieve `compile_commands.json`:

```
scp -P 2022 root@localhost:/paddle/build/compile_commands.json .
```

An example usage of the `compile_commands.json` is to understand how `cmake` built a specific pacakge, e.g., `paddle/cuda`.  To do this, we just need to have a look at JSON records has property `directory: /paddle/build/paddle/cuda`.
